### PR TITLE
[FEATURE] Add KinematicContactProbe sensor.

### PR DESCRIPTION
**Pull request by Milotrince**
## Description
This PR replaces https://github.com/Genesis-Embodied-AI/Genesis/pull/2292

I implement KinematicContactProbe as:
1. `solver.collider.get_contacts()`
2. Filter contacts by links the probes are attached to
3. For each probe, raycast against geoms in contact in the direction of -probe_normal
image

4. For probes with no rayhit, compute collision of probe sphere vs geoms, use distance to contact point projected along normal as penetration depth
image

## Related Issue
Resolves https://github.com/Genesis-Embodied-AI/Genesis/issues/2254

## Motivation and Context
With contact information from the solver, we get contacts only at support points.
The kinematic probes should be able to get information about "contact" at arbitrary points.

## How Has This Been / Can This Be Tested?
updated `test_sensors.py`

## Screenshots (if appropriate):

https://github.com/user-attachments/assets/e18bbf79-0c1b-42be-8b69-b93fa30159bf

## Checklist:
- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [ ] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [x] I tested my changes and added instructions on how to test it for reviewers.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: tests/test_sensors.py [hunk 1]
```diff
@@ -627,3 +627,112 @@ def test_lidar_cache_offset_parallel_env(show_viewer, tol):
...
+    assert_allclose(sphere_data.normal, sphere_ground_truth.normal, tol=gs.EPS)
+    assert_allclose(sphere_data.force, sphere_ground_truth.force, tol=gs.EPS)
+
+    zeros_3 = (0.0, 0.0, 0.0)
+    assert_allclose(true_data.penetration, (PENETRATION, 0.0, 0.0), tol=tol)
+    assert_allclose(true_data.position, ((0.0, 0.0, BOX_SIZE / 2 - PENETRATION), zeros_3, zeros_3), tol=tol)
+    assert_allclose(true_data.normal, ((0.0, 0.0, 1.0), zeros_3, zeros_3), tol=tol)
+    assert_allclose(true_data.force, ((0.0, 0.0, STIFFNESS * PENETRATION), zeros_3, zeros_3), tol=tol)
+    # TODO: This is failing, probably due to how the sensor uses support points?
+    assert_allclose(sphere_data.penetration, PENETRATION, tol=tol)
```
Comment (Milotrince) [inline tests/test_sensors.py]: will un-draft the PR once I address this

File: genesis/engine/solvers/rigid/collider/narrowphase.py [hunk 2]
```diff
@@ -47,6 +43,19 @@ class CCD_ALGORITHM_CODE(IntEnum):
...
+def func_point_in_geom_aabb(
+    i_g: ti.i32,
+    i_b: ti.i32,
+    geoms_state: array_class.GeomsState,
+    point: ti.types.vector(3, gs.ti_float),
+    expansion: gs.ti_float = 0.0,
+):
+    aabb_min = geoms_state.aabb_min[i_g, i_b] - expansion
+    aabb_max = geoms_state.aabb_max[i_g, i_b] + expansion
+    return (point >= aabb_min).all() and (point <= aabb_max).all()
```
Comment (YilingQiao) [inline genesis/engine/solvers/rigid/collider/narrowphase.py]: you changed >< to >=, <=, is this intentional?

File: genesis/engine/sensors/kinematic_contact_probe.py [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+@register_sensor(KinematicContactProbeOptions, KinematicContactProbeMetadata, KinematicContactProbeData)
+@ti.data_oriented
+class KinematicContactProbe(
+    RigidSensorMixin[KinematicContactProbeMetadata],
+    NoisySensorMixin[KinematicContactProbeMetadata],
+    Sensor[KinematicContactProbeMetadata],
+):
+    """Kinematic contact probe measuring penetration depth along the probe normal on collisions."""
+
+    _update_on_read: bool = True
```
Comment (YilingQiao) [inline genesis/engine/sensors/kinematic_contact_probe.py]: it's not used

File: genesis/engine/sensors/kinematic_contact_probe.py [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+            for side in ti.static(range(2)):
+                contact_link = c_link_a if side == 0 else c_link_b
+                i_g = c_geom_b if side == 0 else c_geom_a
+
+                # Is this contact relevant to this sensor?
+                if contact_link == sensor_link_idx and func_point_in_geom_aabb(
+                    i_g, i_b, geoms_state, probe_pos, radius
+                ):
+                    # Update vertices only for geoms that passed filtering
+                    func_update_verts_for_geom(
```
Comment (YilingQiao) [inline genesis/engine/sensors/kinematic_contact_probe.py]: Potential data race. Different probes can call this function from different threads. Not sure if it will cause bug, but it's certainly expensive? Will it be more efficient if you update verts for all geoms when using this sensor?

File: genesis/options/sensors/options.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
 from typing import TYPE_CHECKING, Sequence
 
 import numpy as np
+import torch
```
Comment (YilingQiao) [inline genesis/options/sensors/options.py]: unused

File: genesis/engine/solvers/rigid/collider/broadphase.py [hunk 6]
```diff
@@ -59,12 +77,7 @@ def func_check_collision_valid(
         i_lb = geoms_info.link_idx[i_gb]
 
         # Filter out collision pairs that are involved in dynamically registered weld equality constraints
-        for i_eq in range(rigid_global_info.n_equalities[None], constraint_state.ti_n_equalities[i_b]):
-            if equalities_info.eq_type[i_eq, i_b] == gs.EQUALITY_TYPE.WELD:
-                i_leqa = equalities_info.eq_obj1id[i_eq, i_b]
-                i_leqb = equalities_info.eq_obj2id[i_eq, i_b]
-                if (i_leqa == i_la and i_leqb == i_lb) or (i_leqa == i_lb and i_leqb == i_la):
-                    is_valid = False
+        is_valid = func_check_weld_constraint(i_la, i_lb, i_b, rigid_global_info, constraint_state, equalities_info)
```
Comment (YilingQiao) [inline genesis/engine/solvers/rigid/collider/broadphase.py]: why modify this?

File: genesis/engine/sensors/kinematic_contact_probe.py [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+    sensor_cache_start: ti.types.ndarray(),
+    sensor_probe_start: ti.types.ndarray(),
+    collider_state: array_class.ColliderState,
+    geoms_state: array_class.GeomsState,
+    geoms_info: array_class.GeomsInfo,
+    fixed_verts_state: array_class.VertsState,
+    free_verts_state: array_class.VertsState,
+    verts_info: array_class.VertsInfo,
+    faces_info: array_class.FacesInfo,
+    default_raycast_depth: ti.f32,
```
Comment (YilingQiao) [inline genesis/engine/sensors/kinematic_contact_probe.py]: gs.ti_float

File: examples/sensors/kinematic_contact_probe.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        print("  [\\]         Reset pusher position")
+    else:
+        print(f"Running headless for {args.seconds}s ...")
+    print()
+
+    # ── Simulation loop ────────────────────────────────────────────────
+    steps = int(args.seconds / scene.sim_options.dt) if not args.vis else None
+    step = 0
+
+    try:
```
Comment (YilingQiao) [inline examples/sensors/kinematic_contact_probe.py:186]: do we need this `try`?

File: genesis/engine/sensors/kinematic_contact_probe.py (hunk 4 above)
Comment (duburcqa) [reply genesis/engine/sensors/kinematic_contact_probe.py]: Race condition is not possible for our Taichi kernels. But definitely useless to call this update multiple times of it is the case.

File: genesis/engine/sensors/kinematic_contact_probe.py (hunk 7 above)
Comment (duburcqa) [reply genesis/engine/sensors/kinematic_contact_probe.py]: Actually we use raw taichi types in signatures at every places for now. This is legacy because in the past Genesis was not guaranteed to be initialized before importing taichi kernels modules.

File: genesis/engine/sensors/kinematic_contact_probe.py (hunk 4 above)
Comment (Milotrince) [reply genesis/engine/sensors/kinematic_contact_probe.py]: `func_update_verts_for_geom` already does a `geoms_state.verts_updated` check. since in most cases there are few geoms in contact, updating all geoms once is probably more expensive than updating the ones we need?

File: genesis/engine/solvers/rigid/collider/broadphase.py (hunk 6 above)
Comment (Milotrince) [reply genesis/engine/solvers/rigid/collider/broadphase.py]: earlier i used this check for the probe kernel but I dont need it now since we have the filter on contacts from the collider. will remove this change

File: examples/sensors/kinematic_contact_probe.py (hunk 8 above)
Comment (Milotrince) [reply examples/sensors/kinematic_contact_probe.py:186]: its not critical but we have this pattern in other examples as well, clean break on keyboard interrupt

File: genesis/engine/sensors/kinematic_contact_probe.py (hunk 4 above)
Comment (duburcqa) [reply genesis/engine/sensors/kinematic_contact_probe.py]: sure, but calling a kernel may be costly even if it does nothing, if the GPU queue is already empty.

Anyway, I'm in favour of starting with something simple and optimising at a later stage, rather than keeping this PR is the pipe forever.

Comment (Milotrince) [reply genesis/engine/sensors/kinematic_contact_probe.py]: we're already in a kernel, this is calling a ti func so that cost is not incurred right?

Comment (duburcqa) [reply genesis/engine/sensors/kinematic_contact_probe.py]: Ah yes indeed! in that case it is already perfect I think.

Comment (duburcqa) [reply genesis/engine/sensors/kinematic_contact_probe.py]: Ok After reading the code more carefully, @YilingQiao is right. there is a race condition issue because you are parallelising over both envs and probes, and probes could be attached to the same geometries. The way it is currently guarded by `verts_updated` is not thread-safe.

Comment (duburcqa) [reply genesis/engine/sensors/kinematic_contact_probe.py]: I think you should do what @YilingQiao suggest and start updating first from all geoms. If you think this may be too costly, then pass a pre-computed list of geometry indices to update as input argument of the kernel. You should be able to parallelise over envs + updated geometries, so it should be very fast.

File: examples/sensors/kinematic_contact_probe.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+            y = (j - centre) * spacing
+            z = platform_height / 2  # top surface in link-local frame
+
+            r = PROBE_RADIUS + i * (PROBE_RADIUS / 10.0)
+
+            positions.append((x, y, z))
+            normals.append((0.0, 0.0, 1.0))
+            radii.append(float(r))
+
+    return positions, normals, radii
```
Comment (duburcqa) [inline examples/sensors/kinematic_contact_probe.py:44]: What about vectoring this?
```python
i = np.repeat(np.arange(grid_n), grid_n)
j = np.tile(np.arange(grid_n), grid_n)
x = (i - centre) * spacing
y = (j - centre) * spacing
z = np.full_like(x, platform_height / 2)  # top surface in link-local frame
positions = np.stack([x, y, z], axis=-1)
normals = np.tile([0.0, 0.0, 1.0], (grid_n * grid_n, 1))
radii = PROBE_RADIUS + i * (PROBE_RADIUS / 10.0)
```

File: examples/sensors/kinematic_contact_probe.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+    parser.add_argument("--cpu", action="store_true", help="Run on CPU instead of GPU")
+    parser.add_argument("--seconds", "-t", type=float, default=3.0, help="Seconds to simulate (headless mode)")
+    args = parser.parse_args()
+
+    gs.init(backend=gs.cpu if args.cpu else gs.gpu, precision="32", logging_level="info")
+
+    scene = gs.Scene(
+        sim_options=gs.options.SimOptions(
+            gravity=(0.0, 0.0, -9.81),
+        ),
```
Comment (duburcqa) [inline examples/sensors/kinematic_contact_probe.py]: Remove this block if useless.

File: examples/sensors/kinematic_contact_probe.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+        )
+    )
+
+    pusher_start = np.array([0.0, 0.0, PLATFORM_HEIGHT + PUSHER_SIZE / 2 - 0.02], dtype=np.float32)
+
+    pusher = scene.add_entity(
+        gs.morphs.Cylinder(
+            radius=PUSHER_SIZE,
+            height=PUSHER_SIZE,
+            pos=tuple(pusher_start),
```
Comment (duburcqa) [inline examples/sensors/kinematic_contact_probe.py]: No need to cast to tuple.

File: genesis/engine/solvers/rigid/collider/utils.py [hunk 12]
```diff
@@ -19,6 +21,19 @@ def func_det3(
...
 
 
+@ti.func
+def func_point_in_geom_aabb(
+    i_g: ti.i32,
+    i_b: ti.i32,
+    geoms_state: array_class.GeomsState,
+    point: ti.types.vector(3, ti.f32),
+    expansion: ti.f32 = 0.0,
+):
```
Comment (duburcqa) [inline genesis/engine/solvers/rigid/collider/utils.py]: It would be nice to "match" the argument order of `func_is_geom_aabbs_overlap` right below.

File: genesis/options/sensors/options.py [hunk 13]
```diff
@@ -181,6 +181,70 @@ def model_post_init(self, _):
             gs.raise_exception(f"resolution must be a float or tuple of 3 floats, got: {self.resolution}")
 
 
+class KinematicContactProbe(RigidSensorOptionsMixin, NoisySensorOptionsMixin, SensorOptions):
+    """
+    Kinematic contact probe for detecting contact without affecting physics simulation.
+
+    The returned "force" is non-physical estimate: F = stiffness * penetration * probe_normal.
```
Comment (duburcqa) [inline genesis/options/sensors/options.py]: Could you elaborate on the meaning of "non-physical estimate". For an uninitiated like me this is very disturbing.

File: genesis/options/sensors/options.py [hunk 14]
```diff
@@ -181,6 +181,70 @@ def model_post_init(self, _):
...
+
+    Note
+    ----
+    If the probe is attached to a fixed entity, it will not detect contacts with other fixed entities.
+
```
Comment (duburcqa) [inline genesis/options/sensors/options.py:193]: Could you add some reference about what is "kinematic contact probe" and how it relates to tactile array? You had a new illustration on your PR / issue. It would be nice to reference it here maybe? Or point to the official documentation and add a section with a video and nice illustration in genesis-doc?

File: genesis/options/sensors/options.py [hunk 15]
```diff
@@ -181,6 +181,70 @@ def model_post_init(self, _):
...
+            gs.raise_exception(f"radius must be non-negative, got: {self.radius}")
+        if self.stiffness < 0:
+            gs.raise_exception(f"stiffness must be non-negative, got: {self.stiffness}")
+
+        probe_local_pos = self._validate_probe_arrays(self.probe_local_pos)
+        probe_local_normal = self._validate_probe_arrays(self.probe_local_normal)
+        norms = np.linalg.norm(probe_local_normal, axis=1)
+        bad_idx = np.where(norms < gs.EPS)[0]
+        if bad_idx.size > 0:
+            idx = int(bad_idx[0])
```
Comment (duburcqa) [inline genesis/options/sensors/options.py]:
```python
is_invalid_mask = norms < gs.EPS
if np.any(is_invalid_mask):
    invalid_idx, _ = np.where(is_invalid_mask)
    gs.raise_exception(f"probe_local_normal[{invalid_idx}] must be non-zero, got: {probe_local_normal[is_invalid_mask]}")
```

File: genesis/engine/sensors/kinematic_contact_probe.py [hunk 16]
```diff
@@ -0,0 +1,445 @@
...
+    stiffness: torch.Tensor = make_tensor_field((0,))
+
+    probe_sensor_idx: torch.Tensor = make_tensor_field((0,), dtype=torch.int32)
+    probe_positions: torch.Tensor = make_tensor_field((0, 3))
+    probe_normals: torch.Tensor = make_tensor_field((0, 3))
+    probe_max_raycast_range: float = 0.0
+
+    n_probes_per_sensor: torch.Tensor = make_tensor_field((0,), dtype=torch.int32)
+    sensor_cache_start: torch.Tensor = make_tensor_field((0,), dtype=torch.int32)
+    sensor_probe_start: torch.Tensor = make_tensor_field((0,), dtype=torch.int32)
```
Comment (duburcqa) [inline genesis/engine/sensors/kinematic_contact_probe.py]: `dtype=gs.tc_int`

File: genesis/engine/sensors/kinematic_contact_probe.py [hunk 17]
```diff
@@ -0,0 +1,445 @@
...
+        # Store n_probes before super().__init__() since _get_return_format() is called there
+        self._n_probes = len(sensor_options.probe_local_pos)
+
+        super().__init__(sensor_options, sensor_idx, data_cls, sensor_manager)
+
+        self._debug_objects: list["Mesh | None"] = []
+        self._probe_local_pos = torch.tensor(self._options.probe_local_pos, dtype=gs.tc_float, device=gs.device)
+        self._probe_local_normal = torch.tensor(self._options.probe_local_normal, dtype=gs.tc_float, device=gs.device)
+        norms = self._probe_local_normal.norm(dim=1, keepdim=True).clamp(min=gs.EPS)
+        self._probe_local_normal /= norms
```
Comment (duburcqa) [inline genesis/engine/sensors/kinematic_contact_probe.py]: `self._probe_local_normal /= self._probe_local_normal.norm(dim=1, keepdim=True).clamp(min=gs.EPS)`

File: examples/sensors/kinematic_contact_probe.py (hunk 11 above)
Comment (Milotrince) [reply examples/sensors/kinematic_contact_probe.py]: I do need to cast to tuple here since np arrays are not of type Sequence. we should add support for pydantic np.ndarray type

Comment (duburcqa) [reply examples/sensors/kinematic_contact_probe.py]: We have strict pydantic now? I see.
